### ISSUE TITLE
Make Horovod Ops Differentiable

### ISSUE THREAD

**Issue by ppwwyyxx** [uber/horovod#318]
The TensorFlow `nccl.all_sum` (and some other nccl ops) is differentiable. Why not make horovod ops differentiable as well? There are use cases.

**Comment by alsrgv** [comment]
Good idea.  Here's [reference implementation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/nccl/python/ops/nccl_ops.py) in TF.

**Comment by tgaddair** [comment]
Have some changes ready for TensorFlow.  Will also add support for PyTorch.

**Comment by tgaddair** [comment]
Hey @ppwwyyxx , did you have a specific use case in mind of this?  @alsrgv and I were just chatting, and we think there might be more than one valid way to compute the gradients here (and possible memory implications) depending on the desired behavior.  Let me know if you have thoughts around what should be computed and what they could be used for.

**Comment by ppwwyyxx** [comment]
For example, in a normalization layer, compute the mean and variance of all the batches across all GPUs, and use it to perform normalization.

**Comment by alsrgv** [comment]
@ppwwyyxx, can you try #331 and let us know if it works well for your use case?

**Comment by ppwwyyxx** [comment]
I implemented cross-gpu batchnorm (https://arxiv.org/abs/1711.07240) with tf.contrib.nccl and horovod allreduce respectively. 
Then I replace all batchnorm in a ResNet50 with CGBN to benchmark the speed on 8 V100s.

Baseline speed with standard batchnorm is similar between horovod and non-horovod (replicated) implementation, about 2500~2600im/s.
CGBN based on horovod is 920im/s. CGBN based on tf.contrib.nccl is 1380im/s.

My implementation of CGBN is https://github.com/tensorpack/tensorpack/blob/07783edb998cec3ec91c4312b39bd754cf9ececa/tensorpack/models/batch_norm.py#L198-L253

**Comment by alsrgv** [comment]
@ppwwyyxx, I think that's a great start!  Do you see the same convergence?  Can you capture [Horovod Timeline](https://github.com/uber/horovod/blob/master/docs/timeline.md) and upload it here, so we could look into improving performance?

**Comment by ppwwyyxx** [comment]
In terms of accuracy I'm seeing expected results up to now but the training is still far from completion. 

The speed is actually not bad for me because the main use case is for detection/segmentation. On mask-rcnn I observed a much smaller performance difference compared to vanilla BN. But of course the faster the better, I'll try to get a timeline.

**Comment by ppwwyyxx** [comment]
[timelines.zip](https://github.com/uber/horovod/files/2147477/timelines.zip)

Two timelines with and without cross-gpu batchnorm, obtained from the command `mpirun -np 8 --output-filename test.log python3 ./imagenet-resnet-horovod.py -d 50 --batch 64 --fake` on https://github.com/tensorpack/benchmarks/tree/master/ResNet-Horovod, with the following diff:
```diff
@@ -31,7 +31,8 @@ class Model(ImageNetModel):
 
     def get_logits(self, image):
         with argscope([Conv2D, MaxPooling, GlobalAvgPooling, BatchNorm], data_format='NCHW'):
-            return resnet_backbone(image, self.num_blocks, resnet_group, resnet_bottleneck)
+            with argscope(BatchNorm, sync_statistics='horovod'):
+                return resnet_backbone(image, self.num_blocks, resnet_group, resnet_bottleneck)
 
     def build_graph(self, *inputs):
         """
@@ -135,8 +136,8 @@ def get_config(model, fake=False):
         model=model,
         data=data,
         callbacks=callbacks,
-        steps_per_epoch=steps_per_epoch,
-        max_epoch=35 if args.fake else 95,
+        steps_per_epoch=50,
+        max_epoch=1,
     )
```

**Comment by tgaddair** [comment]
This is awesome!  Thanks for running these benchmarks, @ppwwyyxx.

**Comment by ppwwyyxx** [comment]
btw, if I remove the allreduce in cross-gpu BN, the performance is 1556im/s. The difference from standard BN is purely due to the use of multiple small kernels rather than cudnn's batchnorm kernel.

**Comment by tgaddair** [comment]
Hey @ppwwyyxx, I was running your tensorpack benchmarks with these changes over the weekend.  Could you tell me how you obtained the images/sec metric?  Was that running a separate training script?  Thanks.

**Comment by ppwwyyxx** [comment]
I computed it myself. Using #images per epoch / time per epoch (in progress bar).
#images per epoch = #steps per epoch (in progress bar) * args.batch * #GPUs.

**Comment by tgaddair** [comment]
Thanks for clarifying @ppwwyyxx.  

Your numbers seem to be a few orders of magnitude better than mine, am I miscalculating something here, or is this amount of variance reasonable?

`mpirun --allow-run-as-root -np 2 --output-filename test.log python3 ./benchmarks/ResNet-Horovod/imagenet-resnet-horovod.py -d 50 --batch 64 --fake`

Output from progress bar:

`100%|##########|50/50[01:11<00:00, 0.70it/s]
[0702 17:34:04 @base.py:252] Epoch 1 (global_step 50) finished, time:1 minute 11 seconds.`

Calculation:

`steps_per_epoch = 50
batch = 64
gpus = 2
time_per_epoch = 71.0
images_per_second = 90.14`

**Comment by ppwwyyxx** [comment]
Could you use the second or later epochs instead? First epoch is always slower.

**Comment by tgaddair** [comment]
Thanks, that brought the epoch time down to 56.3 seconds, for about `114 im/s`.  Still seems low, but perhaps it's a hardware limitation.  I'll run a few more tests with and without the Horovod gradients and see how it goes.

From your standpoint, does everything look reasonable enough to warrant landing these changes (performance, convergence, etc.)?

**Comment by tgaddair** [comment]
These changes have been merged into master, and a new release is being made.  So closing this issue, but let me know if anything else comes up.  And thanks @ppwwyyxx!

**Comment by iliadsouti** [comment]
Hi,
I have a question about CGBN. In the code https://github.com/tensorpack/tensorpack/blob/07783edb998cec3ec91c4312b39bd754cf9ececa/tensorpack/models/batch_norm.py#L198-L253, it is mentioned that the statistics are computed per-machine (multiple GPU). Can it be done across all the workers on different machines?

**Comment by ppwwyyxx** [comment]
@iliadsouti It says per-machine for the nccl implementation. It's across workers for the horovod implementation. The docs were updated.

**Comment by Oktai15** [comment]
@ppwwyyxx can I use your batchnorm without tensorpack trainer and tensorpack at all?

**Comment by ppwwyyxx** [comment]
https://tensorpack.readthedocs.io/modules/models.html#tensorpack.models.BatchNorm
The docs say you cannot.
